### ISSUE TITLE
Crash on warn_for_todos

### ISSUE THREAD

**Issue by kylejm** [hanneskaeufler/danger-todoist#22]
Not sure where the problem actually lies here. Perhaps you can confirm to me (my Ruby knowledge is quite limited 😅).

Using danger-todoist 1.0.1.

How can I help you with debugging?

```
/usr/local/lib/ruby/gems/2.3.0/gems/git-1.3.0/lib/git/diff.rb:64:in `[]'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/lib/danger/danger_core/plugins/dangerfile_git_plugin.rb:115:in `diff_for_file'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.0.1/lib/todoist/plugin.rb:117:in `block in diffs_of_interest'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.0.1/lib/todoist/plugin.rb:117:in `map'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.0.1/lib/todoist/plugin.rb:117:in `diffs_of_interest'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.0.1/lib/todoist/plugin.rb:103:in `find_todos'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.0.1/lib/todoist/plugin.rb:95:in `call_method_for_todos'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.0.1/lib/todoist/plugin.rb:47:in `warn_for_todos'
  Dangerfile:20:in `block in parse'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/lib/danger/danger_core/dangerfile.rb:195:in `eval'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/lib/danger/danger_core/dangerfile.rb:195:in `block in parse'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/lib/danger/danger_core/dangerfile.rb:191:in `instance_eval'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/lib/danger/danger_core/dangerfile.rb:191:in `parse'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/lib/danger/danger_core/executor.rb:47:in `run'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/lib/danger/commands/runner.rb:56:in `run'
  /usr/local/lib/ruby/gems/2.3.0/gems/claide-1.0.0/lib/claide/command.rb:334:in `run'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/danger-b3f452417f0d/bin/danger:5:in `<top (required)>'
  /usr/local/lib/ruby/gems/2.3.0/bin/danger:23:in `load'
  /usr/local/lib/ruby/gems/2.3.0/bin/danger:23:in `<top (required)>'
```

**Comment by hanneskaeufler** [comment]
Whoah, so the stacktrace points to [ruby-git](https://github.com/schacon/ruby-git/blob/master/lib/git/diff.rb#L64) blowing up. Because that's a heavily used gem, I suspect I'm rather using it wrong than this being a bug in ruby-git. Maybe we are passing in something weird to the `diff_for_file` method [here](https://github.com/hanneskaeufler/danger-todoist/blob/master/lib/todoist/plugin.rb#L117). Let me play around a bit, maybe this happens when renaming a file? Do you know whats in the diff for the PR (modifications, renames, deletions)? Thanks for reporting!

**Comment by kylejm** [comment]
Sorry for slow response. There were renames, modifications, deletions and additions. So everything! 😅.

Something unusual about our repo is that our default branch is our `develop` branch, not `master`. So `master` is behind `develop` and is representing the live version of the our iOS app, `develop` represents the latest changes that are ready to go into a release candidate. Don't know if that sheds any light for comparisons? Likely not, but I've seen problems arise in Danger because of this setup.

**Comment by kylejm** [comment]
I'll try commenting todoist usage, merge, and then test the PR with `danger local` with todoist uncommented so see if I can learn more.

**Comment by hanneskaeufler** [comment]
I couldn't reproduce the bug so far, deletions/renames should not affect anything as I think about it. Yeah, maybe the branching trips something up, but obviously this has to be fixed ❗ Any further info would be great, thanks for reporting!

**Comment by kylejm** [comment]
Just to update you, I haven't forgotten about this. I'm in a really busy period at my company as we approach a major release. Keeping this on my TODO (😉) list.

**Comment by kylejm** [comment]
After a couple of months, I've finally got around to this one 😅. Sorry for the delay!

It looks like this issue was being caused by an issue with Danger when the repo was cloned on the CI with a shallow depth (which we do). This was [fixed a month or ago](https://github.com/danger/danger/pull/636).

I've re-enabled danger-todoist for us, I'll update if I see any issues :). Assume there are none if I don't re-open though.

**Comment by hanneskaeufler** [comment]
Epic, thanks for getting back to this and explaining the issue, appreciate it!

**Comment by kylejm** [comment]
Bad news... It looks like the issue is indeed in todoist :(. Here's the stack trace:

```
[!] The exception involves the following plugins:
 -  danger-todoist
bundler: failed to load command: danger (/usr/local/bin/danger)
Danger::DSLError: 
[!] Invalid `Dangerfile` file: undefined method `[]' for nil:NilClass
 #  from Dangerfile:21
 #  -------------------------------------------
 #  
 >  todoist.warn_for_todos
 #  todoist.print_todos_table
 #  -------------------------------------------

  /usr/local/lib/ruby/gems/2.3.0/gems/git-1.3.0/lib/git/diff.rb:64:in `[]'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/danger_core/plugins/dangerfile_git_plugin.rb:115:in `diff_for_file'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.1.0/lib/todoist/plugin.rb:117:in `block in diffs_of_interest'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.1.0/lib/todoist/plugin.rb:117:in `map'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.1.0/lib/todoist/plugin.rb:117:in `diffs_of_interest'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.1.0/lib/todoist/plugin.rb:103:in `find_todos'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.1.0/lib/todoist/plugin.rb:95:in `call_method_for_todos'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-todoist-1.1.0/lib/todoist/plugin.rb:47:in `warn_for_todos'
  Dangerfile:21:in `block in parse'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/danger_core/dangerfile.rb:199:in `eval'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/danger_core/dangerfile.rb:199:in `block in parse'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/danger_core/dangerfile.rb:195:in `instance_eval'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/danger_core/dangerfile.rb:195:in `parse'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/danger_core/dangerfile.rb:272:in `run'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/danger_core/executor.rb:27:in `run'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/lib/danger/commands/runner.rb:66:in `run'
  /usr/local/lib/ruby/gems/2.3.0/gems/claide-1.0.1/lib/claide/command.rb:334:in `run'
  /usr/local/lib/ruby/gems/2.3.0/gems/danger-4.0.1/bin/danger:5:in `<top (required)>'
  /usr/local/bin/danger:23:in `load'
  /usr/local/bin/danger:23:in `<top (required)>'
```

Let me know if there's anything I can do to help. We're mid-launch for v2 of our product right now, so I may not be that helpful, but I'll do my best :)

**Comment by hanneskaeufler** [comment]
```shell
  1) Danger::DangerTodoist with Dangerfile throws a meaningful error if git gem returns somthing weird
     Failure/Error: expect { @todoist.warn_for_todos }.to_not raise_error

       expected no Exception, got #<NoMethodError: undefined method `[]' for nil:NilClass> with backtrace:
         # /Users/hanneskaeufler/.rvm/gems/ruby-2.2.3/gems/danger-3.4.1/lib/danger/danger_core/plugins/dangerfile_git_plugin.rb:115:in `diff_for_file'
         # ./lib/todoist/plugin.rb:117:in `block in diffs_of_interest'
         # ./lib/todoist/plugin.rb:117:in `map'
         # ./lib/todoist/plugin.rb:117:in `diffs_of_interest'
         # ./lib/todoist/plugin.rb:103:in `find_todos'
         # ./lib/todoist/plugin.rb:95:in `call_method_for_todos'
         # ./lib/todoist/plugin.rb:47:in `warn_for_todos'
         # ./spec/todoist_spec.rb:130:in `block (4 levels) in <module:Danger>'
         # ./spec/todoist_spec.rb:130:in `block (3 levels) in <module:Danger>'
     # ./spec/todoist_spec.rb:130:in `block (3 levels) in <module:Danger>'
```

**Comment by kylejm** [comment]
@cieslakdawid could you take a look at this when you have a moment to help @hanneskaeufler solve the problem?

**Comment by hanneskaeufler** [comment]
#24 should resolve this. I just released v1.2.0 containing this fix. hope this solves you issues. if not feel free to reopen :)

**Comment by kylejm** [comment]
Excellent! I'll get the team to re-enable and test :) 🎉
